django/db/models/fields/json.py

```diff
@@ -0,0 +1,422 @@
+
+    def select_format(self, compiler, sql, params):
+        if compiler.connection.vendor == 'postgresql':
+            # Avoid psycopg2's automatic decoding to allow custom decoder
+            return '%s::text' % sql, params
+        return super().select_format(compiler, sql, params)
+
+    def validate(self, value, model_instance):
+        super().validate(value, model_instance)
+        try:
+            json.dumps(value, cls=self.encoder)
+        except TypeError:
+            raise exceptions.ValidationError(
+                self.error_messages['invalid'],
+                code='invalid',
+                params={'value': value},
+            )
+
+    def value_to_string(self, obj):
+        return self.value_from_object(obj)
+
+    def formfield(self, **kwargs):
+        return super().formfield(**{
+            'form_class': forms.JSONField,
+            'encoder': self.encoder,
+            'decoder': self.decoder,
+            **kwargs,
+        })
+
+
+class JSONLookup(FieldGetDbPrepValueMixin, Lookup):
+    def as_postgresql(self, compiler, connection):
+        lhs, lhs_params = self.process_lhs(compiler, connection)
+        rhs, rhs_params = self.process_rhs(compiler, connection)
+        params = lhs_params + rhs_params
+        return '%s %s %s' % (lhs, self.postgresql_operator, rhs), params
+
+    def as_sql(self, compiler, connection):
+        raise NotSupportedError(
+            _('%s lookup is not supported by this database backend.' % self.lookup_name)
+        )
+
+
+class HasKeyMixin(JSONLookup):
+    mysql_template = "JSON_CONTAINS_PATH({}, '%s', {})"
+    oracle_template = "JSON_EXISTS({}, '{}')"
+    sqlite_template = "JSON_TYPE({}, %s) IS NOT NULL"
+    _one_or_all = 'one'
+    _logical_operator = ''
+
+    def _process_paths(self, compiler, connection):
+        lhs, lhs_params = self.process_lhs(compiler, connection)
+        if isinstance(self.rhs, str):
+            self.rhs = [self.rhs]
+        paths = [
+            '$.{}'.format(json.dumps(key_name))
+            for key_name in self.rhs
+        ]
+        return lhs, lhs_params, paths
+
+    def as_mysql(self, compiler, connection):
+        lhs, lhs_params, paths = self._process_paths(compiler, connection)
+        sql = (self.mysql_template % self._one_or_all).format(lhs, ', '.join('%s' for _ in paths))
+        return sql, lhs_params + paths
+
+    def as_oracle(self, compiler, connection):
+        lhs, lhs_params, paths = self._process_paths(compiler, connection)
+        template = (self.oracle_template.format(lhs, path) for path in paths)
+        sql = '(%s)' % self._logical_operator.join(template)
+        return sql, lhs_params
+
+    def as_sqlite(self, compiler, connection):
+        lhs, lhs_params, paths = self._process_paths(compiler, connection)
+        template = (self.sqlite_template.format(lhs) for _ in paths)
+        sql = '(%s)' % self._logical_operator.join(template)
+        return sql, lhs_params + paths
```

Comment:
IMO `HasKey` lookups have mixed logic, class variables make them hard to understand (at least for me). I would add `quote_key()` hook and try to reuse a common logic, e.g. :
```python
class HasKeyLookup(SimpleOperatorMixin, Lookup):
    logical_operator = None

    def as_sql(self, compiler, connection, template=None):
        lhs, lhs_params = self.process_lhs(compiler, connection)
        rhs = [self.rhs] if not isinstance(self.rhs, (list, tuple)) else self.rhs
        rhs_params = ['$.%s' % json.dumps(key) for key in rhs]
        sql = template % lhs
        if self.logical_operator:
            # Add condition for each key.
            sql = self.logical_operator.join([sql] * len(rhs_params))
        return sql, lhs_params + rhs_params

    def as_oracle(self, compiler, connection):
        sql, params = self.as_sql(compiler, connection, template="JSON_EXISTS(%s, '%%s')")
        # Add paths directly into SQL because path expressions cannot be passed
        # as a bind variables on Oracle.
        return sql % tuple(params), []

    def as_sqlite(self, compiler, connection):
        return self.as_sql(compiler, connection, template='JSON_TYPE(%s, %%s) IS NOT NULL')

    def as_mysql(self, compiler, connection):
        return self.as_sql(compiler, connection, template="JSON_CONTAINS_PATH(%s, 'one', %%s)")


@JSONField.register_lookup
class HasKey(HasKeyLookup):
    lookup_name = 'has_key'
    postgres_operator = '?'
    prepare_rhs = False


@JSONField.register_lookup
class HasAnyKeys(HasKeyLookup):
    lookup_name = 'has_any_keys'
    postgres_operator = '?|'
    logical_operator = ' OR '

    def get_prep_lookup(self):
        return [str(item) for item in self.rhs]


@JSONField.register_lookup
class HasKeys(HasAnyKeys):
    lookup_name = 'has_keys'
    postgres_operator = '?&'
    logical_operator = ' AND '

    def as_mysql(self, compiler, connection):
        return self.as_sql(compiler, connection, template="JSON_CONTAINS_PATH(%s, 'all', %%s)")
```
It's not perfect, but it took me a while and I'm quite happy with that proposition. I hope that it will also help you with other lookups.

Reply:
That looks much better. Thanks a lot!